### PR TITLE
Add Instrumentation support

### PR DESCRIPTION
**Pull request by dwo**
fao @VivienBarousse 

Adds instrumentation support via the ActiveSupport::Notifications API (with some internal non-backwards-breaking refactorings to support this better with all the existing adapters).

### CODE REVIEW

File: lib/songkick/transport/base.rb [hunk 1]
```diff
@@ -48,29 +46,75 @@ def with_params(params)
...
+
       private
 
       def process(url, status, headers, body)
         Response.process(url, status, headers, body, @user_error_codes)
       end
 
+      def instrument(request)
+        if self.instrumenter
+          payload = { :adapter => self.class.name,
```
Comment (dwo) [inline lib/songkick/transport/base.rb]: should I try to match the params used in Reporting better here?

File: lib/songkick/transport/curb.rb [hunk 2]
```diff
@@ -63,6 +61,11 @@ def execute_request(req)
           connection.http(req.verb.upcase)
         end
 
+        if self.instrumenter
+          self.instrumentation_payload_extras[:connect_time] = connection.connect_time
```
Comment (dwo) [inline lib/songkick/transport/curb.rb]: `Curl::Easy` also has its own notion of `total_time`, wasn't sure whether this should be put into the payload. Should I prefix these like `curl_connect_time`, `curl_name_lookup...` etc?

Comment (VivienBarousse) [inline lib/songkick/transport/curb.rb]: I think having them non-prefixed might be better. You could instrument Curb and HTTParty in the same way, without having to worry about the actual implementation in the instrumentation, only worrying about whether the attribute is there or not?

Comment (dwo) [inline lib/songkick/transport/curb.rb]: makes sense

**Pull request merged**